### DISCUSSION TITLE
Concurrency across multiple connectors

### DISCUSSION THREAD

**Discussion by henninghorn** [saloonphp/saloon#289] [Q&A]
Hi!

First of all, Saloon is such a joy to work with!

I'm working with multiple API's and at one point I need to check availability across different hotels using different connectors.
E.g. hotel A and B uses API A, but hotel C and D, uses software B and C.
So a total of four requests across three different API's.

I've read https://docs.saloon.dev/digging-deeper/concurrency-and-pools, but it doesn't seem to be possible to create a pool of requests across multiple connectors.

Can it be achieved in some way?

Best regards
- Henning

**Comment by Sammyjo20** [comment]
Hey @henninghorn thank you for the kind words! ❤️

So this is something that I've actually been curious about for a while myself. It all depends on how Guzzle handles the concurrent requests internally. I **believe that when you send an asynchronous request, it uses the ["cURL Multi Handler" in Guzzle](https://github.com/guzzle/guzzle/blob/7.8/src/Handler/CurlMultiHandler.php), although I may be wrong here. I think this handler keeps the connections open allowing you to make multiple requests on the same connection, which is how you get the speed increase.

When I built the Pooling functionality in Saloon I first understood how it works, and it turns out that the "concurrency" part is mainly handled by Guzzle's promise library - it can invoke multiple promises at once. There must be some magic going on behind the scenes because I'm not sure how the promise library interacts with the HTTP client. This would definitely be something worth looking into!

However, to test your theory - Saloon's pools accept either a request or a `PromiseInterface`. Would you be able to test the following code with your connectors to see if you get concurrency across multiple domains?

```php
// As the pool requires a connector you have to provide it with a connector like this.

$pool = $hotelOneConnector->pool([
    // Then use your other connector to create the promises
   $hotelOneConnector->sendAsync(new SearchRoomAvailability),
   $hotelTwoConnector->sendAsync(new SearchRoomAvailability),
   $hotelThreeConnector->sendAsync(new SearchRoomAvailability),
]);
```

You can set your concurrency as normal. 

Let me know how that goes!

**Reply by henninghorn** [reply]
Hi @Sammyjo20 

I tried your approach:

```php
$pool = $hotelOne->pool(
    requests: [
        $hotelOne->sendAsync(new \App\Http\Integrations\HotelOne\Requests\GetAvailability),
        $hotelTwo->sendAsync(new \App\Http\Integrations\HotelTwo\Requests\GetAvailability),
        $hotelThree->sendAsync(new \App\Http\Integrations\HotelThree\Requests\GetAvailability),
    ],
    concurrency: 5,
);

$pool->withResponseHandler(function (Response $response) {
    dump($response->json());
});

$promise = $pool->send();
$promise->wait();
```

Each request is hitting a test endpoint which just sleeps for a couple of seconds:
![image](https://github.com/saloonphp/saloon/assets/742607/eb0bf2d3-a2a6-492c-9c61-90b6ddefdeba)

But as you can see, the requests are unfortunately not done in parallel.
I have tried with the default http sender and with Laravel's HTTP client, but the results are the same.

I'll let you know if I figure something else out, I'm sure it can be achieved :)

**Reply by juse-less** [reply]
@henninghorn Hi!
Given the Senders you've tried I'm assuming you're using Guzzle with cURL as an HTTP client.

What you could try is to add a public `withSender(Sender $sender)` method on your Connectors.
Then you could try something like
```php
$hotelOne = new HotelOne;
$hotelTwo = new HotelTwo;
$hotelThree = new HotelThree;

$hotelTwo->withSender($hotelOne->getSender());
$hotelThree->withSender($hotelOne->getSender());

$pool = $hotelOne->pool(
    requests: [
        $hotelOne->sendAsync(new \App\Http\Integrations\HotelOne\Requests\GetAvailability),
        $hotelTwo->sendAsync(new \App\Http\Integrations\HotelTwo\Requests\GetAvailability),
        $hotelThree->sendAsync(new \App\Http\Integrations\HotelThree\Requests\GetAvailability),
    ],
    concurrency: 5,
);

$promise = $pool->send();
$promise->wait();
```
